### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.ldg*`/`T.stg*`/`T.lds*` on a `T.StridedTensor` address the wrong element — offset uses buffer shape, not the declared strides

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#3025]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.13 (latest release)

### System information

- TileLang 0.1.13, Python 3.13, Linux.
- GPU: NVIDIA L40S (compute capability 8.9, sm_89). The defect is in the pointer-offset helper `tilelang/utils/language.py` (Python frontend), which is architecture-independent; the wrong offset is baked into the emitted address arithmetic on every target.

### Problem description

`T.StridedTensor((M, N), (PITCH, 1))` declares a pitched/padded global buffer whose physical row stride `PITCH` is larger than the logical width `N`. When such a buffer feeds any of the explicit `retrieve_ptr`-backed load/store primitives (`T.ldg32/64/128/256`, `T.stg32/64/128/256`, `T.lds32/64/128`, `T.ds_read_*`), the flat element offset for `[row, col]` is computed from the buffer's **shape** (row-major `row*N`) rather than its **declared strides** (`row*PITCH`), so every non-first row is accessed at the wrong address. The kernel compiles and runs with no error; the result is silently wrong (and, when the physical buffer is exactly the logical size, the accesses land out of bounds).

With `M,N,PITCH = 4,3,8`, a `T.stg32(dst[r,0], r+1)` loop over a `StridedTensor` writes value `r+1` to flat element `r*3` instead of `r*8`; reading back column 0 gives `[1, 0, 0, 0]` (rows 1–3 scattered into the padding gap) instead of `[1, 2, 3, 4]`. The identical kernel on a contiguous `T.Tensor` gives `[1, 2, 3, 4]`, so this is specific to the strided buffer, not to `T.stg32`. The symmetric load path `T.ldg32` is wrong the same way (on a pitched source whose col-0 rows hold `[10, 18, 26, 34]`, `T.ldg32(src[r,0])` reads `[10, 13, 16, 19]` — i.e. `src[r*3]` — instead).

Not a regression — see Provenance.

### Reproducible example code

```python
import torch, tilelang, tilelang.language as T

M, N, PITCH = 4, 3, 8  # logical (M,N); physical row pitch PITCH > N (padded rows)

def kernel(strided):
    dt = (T.StridedTensor((M, N), (PITCH, 1), "int32") if strided
          else T.Tensor((M, N), "int32"))
    @T.prim_func
    def main(dst: dt):
        with T.Kernel(1, threads=1):
            for r in T.serial(M):
                T.stg32(dst[r, 0], T.Cast("uint32", r + 1))   # write r+1 into element [r, 0]
    return main

def run(strided):
    k = tilelang.compile(kernel(strided), target="cuda")
    if strided:
        phys = torch.zeros(M * PITCH, dtype=torch.int32, device="cuda")
        dst = phys.as_strided((M, N), (PITCH, 1))   # pitched view, stride[0] = PITCH
    else:
        dst = torch.zeros(M, N, dtype=torch.int32, device="cuda")
    k(dst)
    return dst[:, 0].tolist()

print("contiguous  col0 =", run(False))   # -> [1, 2, 3, 4]  (correct control)
print("strided     col0 =", run(True))    # -> [1, 0, 0, 0]  (wrong: rows 1-3 lost)
```

<details><summary>Emitted CUDA and the offset math (the address uses <code>r*3</code>, i.e. shape, not the declared stride <code>8</code>)</summary>

The strided kernel emits the store address with the shape-derived stride `3`, not the declared `PITCH = 8`:

```
tl::store_global_32((&(dst[(r * 3)])), ...);   // should be dst[(r * 8)]
```

The offset comes straight from `retrieve_ptr`: for `dst[2, 0]` on this buffer it accumulates `offset = 2*3 = 6`, where the correct offset is `2*PITCH = 16`. `buffer.access_ptr(offset=...)` treats `offset` as a flat element offset, so a stride computed from shape produces a wrong address directly.

</details>

### Traceback

No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic. (When the physical allocation equals the logical `M*N`, the wrong offsets also read/write out of bounds.)

### Expected behavior

For a `StridedTensor`, the element offset should use the buffer's declared strides (`row*PITCH`), so that `T.stg32`/`T.ldg32` address the same physical layout the buffer describes — matching the contiguous case, which is correct, and matching the region-copy path (`T.copy`, TMA), which threads the buffer's real strides through the TIR buffer rather than recomputing them.

### Additional context

**Root cause.** `retrieve_stride` computes row-major strides from the buffer's `shape` and never reads the buffer's declared `strides`, so any buffer whose real strides differ from its shape (a `StridedTensor`) gets wrong offsets. [`retrieve_stride`](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/tilelang/utils/language.py#L253-L271) loops `for s in reversed(shape): strides.insert(0, stride); stride *= s` — the docstring even states it derives strides "based on its buffer.shape". Its result feeds the offset accumulation in [`retrieve_ptr`](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/tilelang/utils/language.py#L303-L332) (`offset += region[i].min * strides[i]`), which backs `T.ldg32/64/128` and `T.stg32/64/128` (`builtin.py`, `ptr = retrieve_ptr(dst/src, ...)`). `retrieve_stride` is also called in [`gemm_op.py`](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/tilelang/language/gemm_op.py#L96-L97) (`A_stride = retrieve_stride(A_region)`), but the in-tree comment there states that value is deprecated and "no longer read in-tree" (GEMM consumes the complete operand `BufferRegion`s), so GEMM is not a demonstrated wrong-code path here — the confirmed consumers are the explicit `ldg*`/`stg*` primitives.

**Suggested fix.** Have `retrieve_stride` return the buffer's declared `strides` when present (`obj.strides` / `obj.buffer.strides`), falling back to the row-major-from-shape computation only when strides are unset. The same `buffer.strides` field is already carried on the TIR buffer that `StridedTensor` builds, so the offset accumulation in `retrieve_ptr` would then match the buffer's actual layout.

**Provenance.** Present since `retrieve_stride` first shipped in [#701](https://github.com/tile-ai/tilelang/pull/701) ("Support mma gemm with stride", merged 2025-08-11) — the original definition already recomputed strides from `reversed(object.shape)` and ignored the buffer's declared strides; [#1200](https://github.com/tile-ai/tilelang/pull/1200) (merged 2025-11-12) only promoted it to the shared `utils/language.py` helper without changing that logic. Not a regression. Reproduced on 0.1.13 (`8001cc4`); the helper is byte-identical on the current `main`, so it is still live there.

**Dedup.** I searched the open and closed tracker (`retrieve_stride`, `StridedTensor`, stride) and found no existing report of this offset defect. Distinct from [#2755](https://github.com/tile-ai/tilelang/issues/2755): that is the NVRTC **host adapter** dropping dynamic-stride parameters at the launch-marshalling boundary, not the device-side offset arithmetic; this is `retrieve_stride` computing the wrong in-kernel address from a static declared stride.

**Reach.** Verified by running the shipped tests this session (0.1.13, L40S sm_89). `StridedTensor` is a documented public buffer type (`tilelang/language/proxy.py`, shown in the `examples/eager_jit` notebooks) and appears in 10 test/example files at v0.1.13 (`test_tilelang_language_subtype.py`, `..._copy.py`, `..._tma_copy.py`, `..._vectorize.py`, `..._eager_jit.py`, `test_tilelang_issue_2292.py`, the two `eager_jit` notebooks, and the NVRTC/cutedsl host tests). I ran the ones that exercise strides:
- `test_tilelang_language_subtype.py -k "stride or noncontig"` → **12 passed** (0.1.13). These store to a `StridedTensor` via the ordinary indexed store `dst[i,j,k] = ...` (e.g. `test_subtype_fp4_dynamic_stride_store`), which lowers through the normal TIR index path that DOES honor the declared strides — so they pass and never touch `retrieve_ptr`. They dodge the bug because they use ordinary `BufferStore`, not `T.stg*`.
- `test_tilelang_language_ldst.py` + `test_tilelang_language_ldg.py` → **15 passed, 4 skipped**. These are the only shipped `stg*`/`ldg*` tests and they dodge the bug for the other reason: every buffer is a contiguous 1-D `T.Tensor` (`T.Tensor[[N], f32]`), so the shape-derived stride equals the real stride and `retrieve_stride` is accidentally correct.

So no shipped test or example ever crosses a `StridedTensor` (declared stride ≠ shape) with a `retrieve_ptr`-backed primitive — the two conditions that jointly trigger the bug — which is why CI is green. Both ingredients are individually common (pitched allocations for alignment; explicit `ldg`/`stg`); the bug fires at their intersection.

**Two-level root.**
- *Source-level root:* `retrieve_stride` (`tilelang/utils/language.py:253-271`) derives strides from `buffer.shape` and never reads `buffer.strides`; its result feeds the offset accumulation in `retrieve_ptr` (L303-347). The dead-but-exported twin `retrive_ptr_from_buffer_region` (L274-301) has the identical `for i, shape in enumerate(reversed(buffer.shape))` logic (same latent defect, no in-tree caller).
- *Operator-level root:* every primitive whose address is built by `retrieve_ptr` — `T.ldg32/64/128/256`, `T.stg32/64/128/256`, `T.lds32/64/128`, `T.ds_read_tr8/tr16_b64` (all in `builtin.py`), plus the mbar pointers in `cluster.py` / `gemm_tcgen05.py` (which carry no meaningful stride, so benign).

**Generalization sweep — every cell RUN this session on 0.1.13 (sm_89), value/emitted-address recorded:**

| axis | cell tested | observed result | same-root? |
|---|---|---|---|
| (repro) | `T.stg32` store, pitch `(8,1)` shape `(4,3)` | col0 = `[1,0,0,0]`; emits `dst[(r * 3)]` (should be `r*8`) | yes |
| related-operator | `T.ldg32` load, same pitched buffer | reads `[10,13,16,19]`, correct `[10,18,26,34]`; emits `src[(r * 3)]` | yes (symmetric `retrieve_ptr`) |
| related-operator | `T.stg64`/`T.ldg64` (64-bit width) | emits `dst[(r * 4)]` for shape `(4,4)` pitch `(8,1)` (should be `r*8`) | yes, width-independent |
| related-source | 3-D `StridedTensor` `(2,3,4)` strides `(40,10,1)` → `T.stg32` | emits `dst[((a * 12) + (b * 4))]` (shape-derived; should be `a*40+b*10`) | yes, multi-dim accumulation |
| related-operator/scope | `T.lds32` from shared `alloc_shared((M,PITCH))` | reads correct `[0,3,6,9]`; emits `smem[(r_1 * 8)]` | shares code, NOT triggerable — `alloc_shared` gives a buffer whose shape == layout, so shape-derived stride is correct; no frontend makes a shared buffer with declared stride ≠ shape |
| similar-logic (CONTROL) | ordinary indexed store `dst[r,0]=r+1` to the SAME pitched buffer | col0 = `[1,2,3,4]` (correct); emits `dst[(r * 8)]` | **distinct path** — ordinary `BufferStore` honors the declared stride; isolates the defect to `retrieve_ptr` |
| related-source (CONTROL) | `T.copy(strided_src, shared)` region path | matches the pitched source exactly | **distinct path** — threads the buffer's real strides through the TIR `BufferRegion`, does not recompute via `retrieve_stride` |

The bug is a **class over the `retrieve_ptr`-backed primitive family** (`ldg*`/`stg*`/`lds*`/`ds_read_*`), not a single-primitive defect: load and store, all widths, and any rank are wrong identically, because they share the one `retrieve_stride` call. The two CONTROL rows (ordinary store, `T.copy`) are the *correct* strided paths and localize the defect precisely to `retrieve_ptr`. `T.lds*` shares the fragile code but cannot be independently triggered because shared allocations always have shape == layout. Feeding a global `StridedTensor` directly as a `T.gemm` operand is rejected by this build (`Unsupported gemm combination, A: global`) and the `retrieve_stride` call in `gemm_op.py` is dead (see Root cause), so GEMM is neither reachable nor a wrong-code path here.

**Impact.** Silent-wrong when it fires and deterministic (the wrong address is fixed at compile time, so a kernel that hits it is wrong the same way every run). The trigger is narrow — it needs a buffer whose declared stride differs from its shape reaching an explicit `ldg*`/`stg*` primitive — but the consequence is broad: every element whose declared stride differs from the shape-derived one is misaddressed, corrupting the whole output region rather than one value, and when the physical allocation is not over-sized the misaddressing is an out-of-bounds access. Fixing it makes `StridedTensor` — the type whose purpose is to describe a non-contiguous layout — actually honor its declared strides across the `retrieve_ptr`-backed consumers.
